### PR TITLE
Enable OAP gRPC SSL transportation

### PR DESCRIPTION
**Pull request by hanahmily**
Porting to OpenSSL to enable SSL transportation. The server private
key is in the format of PCKS#8, certificates is x509 though.

Signed-off-by: Gao Hongtao <hanahmily@gmail.com>

### CODE REVIEW

File: docs/en/setup/backend/grpc-ssl.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#Support gRPC SSL transportation for OAP server
```
Comment (kezhenxu94) [inline docs/en/setup/backend/grpc-ssl.md]:
```suggestion
# Support gRPC SSL transportation for OAP server
```

File: docs/en/setup/backend/grpc-ssl.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+For OAP communication we are currently using gRPC, a multi-platform RPC framework that uses protocol buffers for
+message serialization. The nice part about gRPC is that it promotes the use of SSL/TLS to authenticate and encrypt
+exchanges. Now OAP support to enable SSL transportation for gRPC receivers.
```
Comment (kezhenxu94) [inline docs/en/setup/backend/grpc-ssl.md]:
```suggestion
exchanges. Now OAP supports to enable SSL transportation for gRPC receivers.
```

File: docs/en/setup/backend/grpc-ssl.md (hunk 1 above)
Comment (kezhenxu94) [reply docs/en/setup/backend/grpc-ssl.md]: without space, the header style is not correctly rendered

**Comment by wu-sheng** [comment]
Please add the ssl into auth test case, single group, e2e

**Comment by wu-sheng** [comment]
@hanahmily This is the agent side TLS doc for you, https://github.com/apache/skywalking/blob/master/docs/en/setup/service-agent/java-agent/TLS.md

**Comment by hanahmily** [comment]
> 1. Only `ReceiverGRPCHandlerRegister` should support only server-side receiver. You could read the codes, only it has the `#addFilter` for `AuthenticationInterceptor`.

Don't get your point.`ReceiverGRPCHandlerRegister` intends to add `interceptor` or `service` to `NettyServerBuilder` by invoking its `addService` method. But SSL should invoke `NettyServerBuilder#sslContext`. I don't find any link between them. Do I miss something? feel free to correct me.

> 2. If we talk about the core module, you have to implement the client side TLS inside `GRPCRemoteClient`, in oder to make sure it matches the core configuration. And this TLS should be used for internal only.

make sense. I should enable client ssl at the same time. There's also a scenario for `sharing-server`, if we want to enable external SSL and disable internal SSL, client ssl needs to read config from `sharing-server` instead of `core`, right?

**Comment by wu-sheng** [comment]
SSL for outside only could active when sharing server activated. Am I right? 
I don't think we should make the internal SSL CA as same as OAP internal. WDYT?

File: docs/en/setup/backend/grpc-ssl.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+## Config OAP server 
+
+You can enable gRPC SSL by add following lines to `application.yml/core/default`.
+```json
+gRPCSslEnabled: true
+gRPCSslKeyPath: /path/to/server-key.pem
+gRPCSslCertChainPath: /path/to/server.crt
+```
+
+If you port to java agent, refer to [TLS.md](../service-agent/java-agent/TLS.md) to config java agent to enable TLS.
```
Comment (wu-sheng) [inline docs/en/setup/backend/grpc-ssl.md]: This linked file should be updated too, to keep consistently

**Comment by hanahmily** [comment]
> SSL for outside only could active when sharing server activated. Am I right?

right

> I don't think we should make the internal SSL CA as same as OAP internal. WDYT?

No. SSL becomes more widely accepted. We should support it both internally and externally.

**Comment by hanahmily** [comment]
@wu-sheng e2e test case is added and passed. now I think we can merge it.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/RemoteClientManager.java [hunk 4]
```diff
@@ -55,21 +59,29 @@
...
      * Initial the manager for all remote communication clients.
-     *
-     * @param moduleDefineHolder for looking up other modules
+     *  @param moduleDefineHolder for looking up other modules
      * @param remoteTimeout      for cluster internal communication, in second unit.
+     * @param grpcSslEnabled true: enable SSL between clusters, false: plain gRPC between the cluster.
+     * @param grpcSslTrustCAPath trust certificate authorities file path.
      */
-    public RemoteClientManager(ModuleDefineHolder moduleDefineHolder, int remoteTimeout) {
+    public RemoteClientManager(ModuleDefineHolder moduleDefineHolder,
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/RemoteClientManager.java]: Could you make the core style consistently. `GRPCRemoteClient` client use two constructors for ssl ON/OFF, here using a boolean and empty/null string.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/GRPCRemoteClient.java [hunk 5]
```diff
@@ -99,7 +110,7 @@ GRPCClient getClient() {
         if (Objects.isNull(client)) {
             synchronized (GRPCRemoteClient.class) {
                 if (Objects.isNull(client)) {
-                    this.client = new GRPCClient(address.getHost(), address.getPort());
+                    this.client = new GRPCClient(address.getHost(), address.getPort(), sslContext);
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/GRPCRemoteClient.java]: A similar code style inconsistent here, using `sslContext==null` represents no SSL.

File: oap-server/server-receiver-plugin/skywalking-sharing-server-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/sharing/server/SharingServerModuleProvider.java [hunk 6]
```diff
@@ -80,6 +81,15 @@ public void prepare() {
         if (config.getGRPCPort() != 0) {
             grpcServer = new GRPCServer(Strings.isBlank(config.getGRPCHost()) ? "0.0.0.0" : config.getGRPCHost(), config
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-sharing-server-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/sharing/server/SharingServerModuleProvider.java]: I think you miss the deletion of this?

File: test/e2e/e2e-ssl/e2e-ssl.iml [hunk 7]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wu-sheng) [inline test/e2e/e2e-ssl/e2e-ssl.iml]: I think this should not be uploaded? Why git ignore is not working for this?

File: test/e2e/e2e-ssl/src/docker/certs/server-key.pem [hunk 8]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (wu-sheng) [inline test/e2e/e2e-ssl/src/docker/certs/server-key.pem]: Are these cert files alway valid? No expired time?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/RemoteClientManager.java (hunk 4 above)
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/remote/client/RemoteClientManager.java]: done. keep all parameter list the same way.

File: oap-server/server-receiver-plugin/skywalking-sharing-server-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/sharing/server/SharingServerModuleProvider.java (hunk 6 above)
Comment (hanahmily) [reply oap-server/server-receiver-plugin/skywalking-sharing-server-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/sharing/server/SharingServerModuleProvider.java]: removed it.

File: test/e2e/e2e-ssl/e2e-ssl.iml (hunk 7 above)
Comment (hanahmily) [reply test/e2e/e2e-ssl/e2e-ssl.iml]: deleted it.

File: test/e2e/e2e-ssl/src/docker/certs/server-key.pem (hunk 8 above)
Comment (hanahmily) [reply test/e2e/e2e-ssl/src/docker/certs/server-key.pem]: 2 years. Is it enough?

Comment (wu-sheng) [reply test/e2e/e2e-ssl/src/docker/certs/server-key.pem]: My point is, this is for test. Could we set it forever? Or let's say 100 years?

**Comment by wu-sheng** [comment]
You need to fix CI

File: test/e2e/e2e-ssl/src/docker/certs/server-key.pem (hunk 8 above)
Comment (hanahmily) [reply test/e2e/e2e-ssl/src/docker/certs/server-key.pem]: Now we get the  100 years/ century certificates.

**Pull request merged**